erigon-lib/downloader/downloader.go

```diff
@@ -673,12 +897,478 @@ func (d *Downloader) mainLoop(silent bool) error {
+	}
+
+	if len(mismatch) > 0 {
+		var webseeds string
+		for _, entry := range mismatch {
+			if len(webseeds) > 0 {
+				webseeds += ", "
+			}
+
+			webseeds += strings.TrimSuffix(entry.url.String(), "/") + "#" + entry.hash.HexString()
+		}
+
+		logger.Warn("Webseed hash mismatch for torrent", "name", name, "hash", torrentHash.HexString(), "webseeds", webseeds)
+	}
+}
+
+func (d *Downloader) checkComplete(name string) bool {
+	if info, err := d.torrentInfo(name); err == nil {
+		if info.Completed != nil && info.Completed.Before(time.Now()) {
+			if info.Length != nil {
+				if fi, err := os.Stat(filepath.Join(d.SnapDir(), name)); err == nil {
+					return fi.Size() == *info.Length
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+func (d *Downloader) getWebDownloadInfo(t *torrent.Torrent) (webDownloadInfo, []*seedHash, error) {
+	torrentHash := t.InfoHash()
+
+	d.lock.RLock()
+	info, ok := d.webDownloadInfo[t.Name()]
+	d.lock.RUnlock()
+
+	if ok {
+		return info, nil, nil
+	}
+
+	seedHashMismatches := make([]*seedHash, 0, len(d.cfg.WebSeedUrls))
+
+	for _, webseed := range d.cfg.WebSeedUrls {
+		downloadUrl := webseed.JoinPath(t.Name())
+
+		if headRequest, err := http.NewRequestWithContext(d.ctx, "HEAD", downloadUrl.String(), nil); err == nil {
+			headResponse, err := http.DefaultClient.Do(headRequest)
+
+			if err != nil {
+				continue
+			}
+
+			headResponse.Body.Close()
+
+			if headResponse.StatusCode == http.StatusOK {
+				if meta, err := getWebpeerTorrentInfo(d.ctx, downloadUrl); err == nil {
+					if bytes.Equal(torrentHash.Bytes(), meta.HashInfoBytes().Bytes()) {
+						// TODO check the torrent's hash matches this hash
+						return webDownloadInfo{
+							url:     downloadUrl,
+							length:  headResponse.ContentLength,
+							torrent: t,
+						}, seedHashMismatches, nil
+					} else {
+						hash := meta.HashInfoBytes()
+						seedHashMismatches = append(seedHashMismatches, &seedHash{url: webseed, hash: &hash})
+						continue
+					}
+				}
+			}
+		}
+
+		seedHashMismatches = append(seedHashMismatches, &seedHash{url: webseed})
+	}
+
+	return webDownloadInfo{}, seedHashMismatches, fmt.Errorf("can't find download info")
+}
+
+func getWebpeerTorrentInfo(ctx context.Context, downloadUrl *url.URL) (*metainfo.MetaInfo, error) {
+	torrentRequest, err := http.NewRequestWithContext(ctx, "GET", downloadUrl.String()+".torrent", nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	torrentResponse, err := http.DefaultClient.Do(torrentRequest)
+
+	if err != nil {
+		return nil, err
+	}
+
+	defer torrentResponse.Body.Close()
+
+	if torrentResponse.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("can't get webpeer torrent unexpected http response: %s", torrentResponse.Status)
+	}
+
+	return metainfo.Load(torrentResponse.Body)
+}
+
+func (d *Downloader) torrentDownload(t *torrent.Torrent, sem *semaphore.Weighted) {
+	t.AllowDataDownload()
```

Comment:
`sem` does limit amount of files to download in same time. So, you need `sem.Acquire` before `t.AllowDataDownload`

Reply:
I've moved the torrent manipulation is to the action go routine so they will happen after the semaphore